Don't always require a native compiler

jon-turney opened the pull request:
Language additions and various fixes so a `meson.build` can be written which doesn't require a native compiler.

`add_languages()` gains a `native:` keyword, indicating if a native or cross compiler is required. 
`project()` gains a`languages:` and `native_languages:` keywords.
Add to CI applicable tests using the MSVC ARM64 cross-compiler only.

Closes: #2392, Closes: #5862, Closes: #5962
Mitigates: #4402. See also #3969

jon-turney commented:
@Ericson2314 I'm particularly uncertain about the last few commits.  Your review of those would be appreciated.

jpakkane commented:
Rather than adding a new kwarg, could we instead try to find both, fail gracefully if only one of them is found and then if the user tries to define a target with a missing language exit with a hard error?

Ericson2314 commented:
I aslso want to split this and first fix `add_language` and only after `project`. @jon-turney if you would be so kind to split the PR like that I'd be happy to help you simplify things. I'll admit I'm surprised and confused by what I think are just `project` changes, but I'm more confident I know how to make `add_languages` simple. After that maybe we'll both together realize how to make `project` simple too :).

jon-turney commented:
> Rather than adding a new kwarg, could we instead try to find both, fail gracefully if only one of them is found and then if the user tries to define a target with a missing language exit with a hard error?

For the avoidance of doubt, I don't thinkthis is possible for `add_languages()`, as that has a return value, which could be amibiguous with lazy detection.

I think even if "'we'" implement that lazy detection, there should also be a syntax for being explicit about which machine(s) languages are used with.

The surface for interacting with compiler detection is not limited to just defining targets, there is also (at least) accessing compiler objects (`meson.get_compiler()`) and the `build|host_machine` objects.

> I also want to split this and first fix `add_language` and only after `project`. @jon-turney if you would be so kind to split the PR like that I'd be happy to help you simplify things.

Good idea.  I split that off as #6532.

jon-turney commented:
Split off #6636 to discuss a machine detection issue.

jon-turney commented:
Refined this in a similar way to add_languages(native:) in #6532, so (existing) `project('whatever', 'c')` means language 'c' is required for the host machine, and detected (but not required) for the build machine.

This is still WIP.  Some test failures are expected and addressed by #6636.

Ericson2314 commented on mesonbuild/ast/introspection.py:
```diff
@@ -124,6 +124,10 @@ def func_project(self, node: BaseNode, args: T.List[TYPE_nvar], kwargs: T.Dict[s
+        for kw in ['languages', 'native_languages']:
+            if kw in kwargs:
+                v = [kwargs.get(kw)]  # type: T.List[TYPE_nvar]
+                self.func_add_languages(None, v, None)
```
`func_add_languages` currently iterates through both `MachineChoice.BUILD`, and `MachineChoice.HOST`. We should give `func_add_languages` a `for_machine` argument instead.

jon-turney replied:
We can't (simply) change the signature of `func_add_languages()`, because it's invoked by the parser function dispatcher.  Certainly an internal `add_languages()` method could be added.

I don't know what use (if any) the introspector makes of the language information.  At this remove in time I no longer know what problem I thought adding this was fixing (reverting it doesn't make any tests fail, now).

Ericson2314 replied:
> Certainly an internal `add_languages()` method could be added.

That sounds good.

> I don't know what use (if any) the introspector makes of the language information.

Even if it does none, I think using a helper function is still good. In that case, it avoids a gotcha going forward if we do try to make it use that info.

jon-turney replied:
I have factored this out as a helper function, `_add_languages()` as you suggest.

Ericson2314 commented on mesonbuild/ast/introspection.py:
```diff
@@ -123,11 +123,12 @@ def func_project(self, node: BaseNode, args: T.List[TYPE_nvar], kwargs: T.Dict[s
-                v = [kwargs.get(kw)]  # type: T.List[TYPE_nvar]
-                self.func_add_languages(None, v, None)
+                self._add_languages([kwargs.get(kw)], MachineChoice.HOST)
+                self._add_languages([kwargs.get(kw)], MachineChoice.BUILD)
```
Ah, I forgot about positional vs keword args. Yeah the logic should be like `interpreter.py`. Right now the new keyword ways is not done altogether.

jon-turney replied:
Done

Ericson2314 commented on mesonbuild/ast/introspection.py:
```diff
@@ -137,17 +142,20 @@ def do_subproject(self, dirname: str) -> None:
 
     def func_add_languages(self, node: BaseNode, args: T.List[TYPE_nvar], kwargs: T.Dict[str, TYPE_nvar]) -> None:
-        args = self.flatten_args(args)
         for for_machine in [MachineChoice.BUILD, MachineChoice.HOST]:
```
And we can do something based off of `kwargs` for this list too? Granted that a pre-existing bug we could also punt for a future PR. I know you've done tons of these revisions already :).

jon-turney replied:
I think I omitted that logic, because as far as I can see, the native kwarg is ignored in the introspector for all other meson functions which support it (so presumably isn't signficant to introspection)

mensinda replied:
Adding `native` support here would be nice, if it would change the introspection output (what meson options printed with `meson configure` without a build dir). The other functions don't use `native` since it wouldn't make any difference to the introspection output (as far as I know).

jon-turney replied:
It seems a bit odd that running the introspector on an unconfigured meson.build will detect compilers, but I guess that makes sense.

jon-turney commented:
Looking at this again, I don't immediately see why the introspector's `_add_lanaguages()` needs to do anything.  Does the introspector make use of the compiler detected anywhere?

mensinda commented:
The introspector needs to know all the supported languages to add the `c_args`, `cpp_args` options, so that the result of `meson configure meson.build` is the same as the output of `meson configure <fresh build dir>` in 99% of all cases.

There are currently no other uses for languages in the introspection code (as far as I can remember).

Ericson2314 commented:
I think all the split-out parts have been merged so this can be rebased?

jon-turney commented:
Updated for CI changes, relies on https://github.com/mesonbuild/cidata/pull/4

Ericson2314 commented:
It seems like there is still a lot of testing-related changes that while great, belongs in a separate PR?

Ericson2314 commented:
The changes to meson itself do look good though!

jon-turney commented:
> It seems like there is still a lot of testing-related changes that while great, belongs in a separate PR?

You'll have to help me understand what parts you have in mind here.

Ericson2314 commented:
@jon-turney everything to do with windows arm cross compilation we could land as a follow-up PR, right?

Ericson2314 commented:
I had just been asking about splitting the PR on principle, but now I have concrete reason to split it up: I'd like to back-port the non-Windows bits to https://github.com/NixOS/nixpkgs/issues/86229 which fix cross builds for Nixpkgs/NixOS after I inadvertently broken them cleaning some things up.

The Windows bits wouldn't hurt, of course, but the smaller the back-port the better.

jon-turney commented:
> @jon-turney everything to do with windows arm cross compilation we could land as a follow-up PR, right?

There is nothing here which is specific to "windows arm cross compilation".

This PR contains a feature "Don't always require a native compiler".  `Contributing.md` says new features require tests.  Building in an environment where only a Windows ARM cross-compiler is available is an instance of that, and is used to test that feature.

> I had just been asking about splitting the PR on principle, but now I have concrete reason to split it up: I'd like to back-port the non-Windows bits

Still have no idea what parts you think can be separately merged.  Please give concrete examples.

Ericson2314 commented:
> Building in an environment where only a Windows ARM cross-compiler is available is an instance of that, and is used to test that feature.

How about test on linux using `CC_FOR_BUILD=garbage` to prevent it from finding a working native C compiler for now? Then later we fix windows cross testing.

> Still have no idea what parts you think can be separately merged. Please give concrete examples.

I'll comment.

Ericson2314 commented on azure-pipelines.yml:
```diff
@@ -53,6 +53,11 @@ jobs:
           arch: x64
           compiler: msvc2019
           backend: vs2019
+        vc2019armninjacross:
+          arch: arm64
+          compiler: msvc2019
+          backend: ninja
+          extraargs: --cross arm64cl.txt
```
Save for windows cross later

Ericson2314 commented on ci/run.ps1:
```diff
@@ -74,4 +74,4 @@ echo "=== Start running tests ==="
 # if it prints anything to stderr. Python's test runner
 # does that by default so we need to forward it.
-cmd /c 'python 2>&1' run_tests.py --backend $env:backend
+cmd /c "python 2>&1 run_tests.py --backend $env:backend $env:extraargs"
```
Save changes in this file for windows cross later

Ericson2314 commented on cross/arm64cl.txt:
```diff
@@ -0,0 +1,14 @@
+[binaries]
+c = 'cl'
+cpp = 'cl'
+ar = 'lib'
+windres = 'rc'
+
+[properties]
+c_args = ['-DWINAPI_FAMILY=WINAPI_FAMILY_APP']
+
+[host_machine]
+system = 'windows'
+cpu_family = 'aarch64'
+cpu = 'armv8'
+endian = 'little'
```
Save for windows cross later

Ericson2314 commented on mesonbuild/environment.py:
```diff
@@ -1094,7 +1094,7 @@ def sanitize(p):
                 return cls(
                     compiler, version, for_machine, is_cross, info, exe_wrap,
-                    target, linker=linker)
+                    target, full_version=cl_signature, linker=linker)
```
Save for windows cross later

Ericson2314 commented on mesonbuild/interpreter.py:
```diff
@@ -2954,8 +2961,19 @@ def func_project(self, node, args, kwargs):
         mlog.log('Project version:', mlog.bold(self.project_version))
-        self.add_languages(proj_langs, True, MachineChoice.BUILD)
-        self.add_languages(proj_langs, True, MachineChoice.HOST)
+
+        if 'languages' in kwargs:
+            if len(proj_langs) > 0:
+                raise InterpreterException('project() may not have both language positional arguments and languages:.')
+            self.add_languages(mesonlib.stringlistify(kwargs.get('languages', [])), True, MachineChoice.HOST)
+        else:
+            self.add_languages(proj_langs, True, MachineChoice.HOST)
+
+        if 'native_languages' in kwargs:
+            self.add_languages(mesonlib.stringlistify(kwargs.get('native_languages', [])), True, MachineChoice.BUILD)
+        else:
+            self.add_languages(proj_langs, False, MachineChoice.BUILD)
+
```
When there are positional languages and `native_languages`, there is no error, but there should be.

How about:
```suggestion

        if len(proj_langs) > 0 and ('native_languages' in kwargs or 'languages' in kwargs):
                raise InterpreterException('project() may not have both language positional arguments and languages:.')
        self.add_languages(proj_langs, True, MachineChoice.BUILD)
        self.add_languages(proj_langs, True, MachineChoice.HOST)
        self.add_languages(mesonlib.stringlistify(kwargs.get('native_languages', [])), True, MachineChoice.BUILD)
        self.add_languages(mesonlib.stringlistify(kwargs.get('languages', [])), True, MachineChoice.HOST)
```

Ericson2314 replied:
Oh nevermind I misread the message

Ericson2314 commented:
Let me just temper the Windows stuff by saying it does look good to me, (and I am excited about it!), and if @jpakkane and others are happy with merging it all at once, it's fine.

jpakkane commented:
The documentation should probably say that most projects should use they keywordless versions as they work 99% of the time. The other versions should only be used for projects with special needs.

Ericson2314 commented:
@jpakkane that would be a bit of friction for Nixpkgs, where we don't give access to native compilers in cross builds unless explicitly requested.

Maybe after a few cycles we could make the list just be for the host platform in cross builds?

jpakkane commented:
Let's look at this through an example. Suppose we have something simple (like zlib) that builds only one library. Its build file should be something like:

```meson
project('foo', 'c')
... = static_library('foo', ...)
... = declare_dependency(...)
``´

This definition **must** work. It **must** work for native building. It **must** work for cross building. It should be the default recommended way of doing things. You should only ever need to specify native/cross compiler requirements _only if_ you build and run code during the build.

The automatic approach of "try to create both compilers, ignore if you can't instantiate a native compiler when cross building but error out when the build file tries to build a native target" is the common solution to the common use case. It makes the common case simple at the cost of making the uncommon one slightly more expensive (specifically by delaying the printing of errors to point of use rather than point of trying to instantiate the compiler). This is a good tradeoff.

Ericson2314 commented:
> This definition must work. It must work for native building. It must work for cross building. It should be the default recommended way of doing things. You should only ever need to specify native/cross compiler requirements only if you build and run code during the build.

We agree.

> try to create both compilers, ignore if you can't instantiate a native compiler when cross building but error out when the build file tries to build a native target

This introduces the same sort of implicit state / non-determinism that environment variables do. We also want users to be able to reproduce their builds, and we want the meson file to say what is needed, not under-approximate it because some things happen to work when extra compilers beyond what is requested are available.

I'm advocating a very small change to your plan, and only after a breaking change warning period:

##### native build
```meson
project('foo', *args)
```
is the same as
```meson
project('foo', languages: args, native_languages: args)
```

##### cross build
```meson
project('foo', *args)
```
is the same as
```meson
project('foo', languages: args, native_languages: [])
```

This means if you don't care about cross, you don't need to care about `native_languages`, even if you build and run code during the build. Only if you do care about cross, *and* use `native: true`, do you need to use `native_languages`.

\[This also means `languages` is no longer necessary.\]

jon-turney commented:
> The documentation should probably say that most projects should use they keywordless versions as they work 99% of the time. The other versions should only be used for projects with special needs.

I've adjusted the documentation to try to address this.

jon-turney commented:
Split out the introspector fix for add_languages(native:) as #7428

nirbheek commented:
@jon-turney would be nice to get this for 0.56, is it possible to split this out into seperate PRs so it's easier to merge?

jon-turney commented:
> @jon-turney would be nice to get this for 0.56, is it possible to split this out into seperate PRs so it's easier to merge?

I'll see what I can do, time permitting.

jon-turney commented:
> How about test on linux using `CC_FOR_BUILD=garbage` to prevent it from finding a working native C compiler for now? Then later we fix windows cross testing.

`CC_FOR_BUILD=garbage` fails with a hard error trying to invoke `garbage`, which is probably what's wanted for `CC_FOR_BUILD=typoed-compiler-command`.

jon-turney commented:
I've dropped the Windows ARM64 cross-only environment testing and replaced that with an arm cross-only environment test.
I've also dropped the additional kwargs for `project()`, since that simplifies things, and was obviously a sticking point.

Ericson2314 reviewed (approved):
Thanks for returning to this! And, of course, after, I'm still excited about that Windows cross stuff.

nirbheek reviewed (approved):
Went through it all, LGTM :)

jon-turney commented:
Split out Windows ARM64 cross-only environment testing as #7718.

jon-turney commented:
With hindsight, moving tests which require a native compiler to a separate suite was probably a mistake.  Now we have `test.json`, we could probably mark tests which require native tools in some way there.  That would also open up the way to cross-testing running suites other than `common`.

nirbheek commented:
Yeah, that sounds like a good idea to me.
